File: apps/pos_app/lib/data/models/sales/promotion_entity.dart
import 'package:floor/floor.dart';

@Entity(tableName: 'promotions')
class PromotionEntity {
  @primaryKey
  final String id;
  final String name;
  final String type; // 'buyXGetYFree', 'percentageDiscount', 'fixedDiscount', 'comboPackage'
  @ColumnInfo(name: 'target_product_id')
  final String? targetProductId;
  @ColumnInfo(name: 'target_category_id')
  final String? targetCategoryId;
  @ColumnInfo(name: 'buy_quantity')
  final int buyQuantity;
  @ColumnInfo(name: 'get_quantity')
  final int getQuantity;
  @ColumnInfo(name: 'discount_value')
  final double discountValue;
  @ColumnInfo(name: 'min_order_amount')
  final double minOrderAmount;
  @ColumnInfo(name: 'days_of_week')
  final String? daysOfWeek; // comma-separated e.g. "5,6"
  @ColumnInfo(name: 'start_time')
  final String? startTime; // "HH:mm"
  @ColumnInfo(name: 'end_time')
  final String? endTime; // "HH:mm"
  @ColumnInfo(name: 'start_date')
  final int? startDate;
  @ColumnInfo(name: 'end_date')
  final int? endDate;
  @ColumnInfo(name: 'priority')
  final int priority;
  @ColumnInfo(name: 'is_stackable')
  final bool isStackable;
  @ColumnInfo(name: 'is_active')
  final bool isActive;

  PromotionEntity({
    required this.id,
    required this.name,
    required this.type,
    this.targetProductId,
    this.targetCategoryId,
    this.buyQuantity = 0,
    this.getQuantity = 0,
    this.discountValue = 0.0,
    this.minOrderAmount = 0.0,
    this.daysOfWeek,
    this.startTime,
    this.endTime,
    this.startDate,
    this.endDate,
    this.priority = 0,
    this.isStackable = true,
    this.isActive = true,
  });
}

File: apps/admin_backend/src/modules/promotions/entities/promotion.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn,
  Index,
} from 'typeorm';
import { Tenant } from '../../tenant/entities/tenant.entity';

export enum PromotionType {
  BUY_X_GET_Y_FREE = 'buyXGetYFree',
  PERCENTAGE_DISCOUNT = 'percentageDiscount',
  FIXED_DISCOUNT = 'fixedDiscount',
  COMBO_PACKAGE = 'comboPackage',
}

@Entity('promotions')
@Index('idx_promotions_tenant_active', ['tenant_id', 'is_active'])
@Index('idx_promotions_tenant_priority', ['tenant_id', 'priority'])
export class Promotion {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  tenant_id: string;

  @ManyToOne(() => Tenant)
  @JoinColumn({ name: 'tenant_id' })
  tenant: Tenant;

  @Column({ type: 'varchar' })
  name: string;

  @Column({
    type: 'enum',
    enum: PromotionType,
    default: PromotionType.BUY_X_GET_Y_FREE,
  })
  type: PromotionType;

  @Column({ type: 'varchar', nullable: true })
  target_product_id?: string | null;

  @Column({ type: 'varchar', nullable: true })
  target_category_id?: string | null;

  @Column({ type: 'int', default: 0 })
  buy_quantity: number;

  @Column({ type: 'int', default: 0 })
  get_quantity: number;

  @Column({
    type: 'numeric',
    precision: 12,
    scale: 2,
    default: 0.0,
    transformer: {
      to: (value: number) => value,
      from: (value: string | number) => (typeof value === 'string' ? parseFloat(value) : value),
    },
  })
  discount_value: number;

  @Column({
    type: 'numeric',
    precision: 12,
    scale: 2,
    default: 0.0,
    transformer: {
      to: (value: number) => value,
      from: (value: string | number) => (typeof value === 'string' ? parseFloat(value) : value),
    },
  })
  min_order_amount: number;

  @Column({ type: 'simple-array', nullable: true })
  days_of_week?: string[] | null; // e.g. ["5", "6"]

  @Column({ type: 'varchar', nullable: true })
  start_time?: string | null; // "17:00"

  @Column({ type: 'varchar', nullable: true })
  end_time?: string | null; // "20:00"

  @Column({ type: 'bigint', nullable: true })
  start_date?: number | null;

  @Column({ type: 'bigint', nullable: true })
  end_date?: number | null;

  @Column({ type: 'int', default: 0 })
  priority: number;

  @Column({ name: 'is_stackable', default: true })
  is_stackable: boolean;

  @Column({ name: 'is_active', default: true })
  is_active: boolean;

  @CreateDateColumn({ name: 'created_at' })
  created_at: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updated_at: Date;
}
